### PR TITLE
vochain: implements deterministic process ID type

### PR DESCRIPTION
**Pull request by p4u**
Introduce a new type `ProcessID` with Marshal and Unmarshal methods.

The new identifier holds information about the process
- chainID: a 6 bytes trunked hash of the blockchain identifier
- organization: hex address of the wallet creating the process
- census proof type: kind of census used for the process (from protobuf models)
- envelope type: flags for configuring the ballot counting mode
- nonce: an incremental uint32 number

So processID: 395b41cc9abcd8da6bf26964af9d7eed9e03e53415d37aa96045062100000001  represents
- chID: vocdoni-azeno
- addr: 0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045
- proof: arbo
- envelopetype: 0x21
- nonce: 1

This commit only introduces the new type, not the required changes on the blockchain logic.

Discussion: https://github.com/vocdoni/protocol/discussions/28

Signed-off-by: p4u <pau@dabax.net>

### CODE REVIEW

**Review by jordipainan — APPROVED** [review]
LGTM!

File: vochain/process_id.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+// chID:vocdoni-azeno addr:0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045 pt:arbo nc:1
+type ProcessID struct {
+	chainID          string
+	organizationAddr *common.Address
+	proofType        uint16
+	nonce            uint32
+}
+
+// Marshal encodes to bytes
+func (p *ProcessID) Marshal() []byte {
```
Comment (mvdan) [inline vochain/process_id.go:37]: might as well implement https://pkg.go.dev/encoding#BinaryMarshaler and its unmarshaler counterpart

File: vochain/process_id.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+//
+// - census proof type: kind of census used for the process (from protobuf models)
+//
+// - nonce: an incremental uint32 number
+//
+// processID: 395b41cc9abcd8da6bf26964af9d7eed9e03e53415d37aa96045000600000001 represents
+// chID:vocdoni-azeno addr:0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045 pt:arbo nc:1
+type ProcessID struct {
+	chainID          string
+	organizationAddr *common.Address
```
Comment (mvdan) [inline vochain/process_id.go]: I don't think you should be using a pointer here. a pointer weighs 8 bytes usually; an address is just 20 bytes.

File: vochain/process_id.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	return p.chainID
+}
+
+// SetAddr sets the organizer address (20 bytes)
+func (p *ProcessID) SetAddr(addr common.Address) {
+	p.organizationAddr = &addr
+}
+
+// Addr returns the organizer address
+func (p *ProcessID) Addr() *common.Address {
```
Comment (mvdan) [inline vochain/process_id.go]: per the above, I don't think this should return a pointer (SetAddr is already good)

File: vochain/process_id.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+		return fmt.Errorf("proof is nil")
+	}
+	// get the Protobuf oneOf descriptor index number and use it as proofType
+	protoDesc := proof.ProtoReflect().Descriptor().Oneofs().Get(0)
+	p.proofType = uint16(proof.ProtoReflect().WhichOneof(protoDesc).Number())
+	return nil
+}
+
+// Proof returns a string representing the census proof type
+func (p *ProcessID) Proof() string {
```
Comment (mvdan) [inline vochain/process_id.go]: nit: SetProof takes a pointer, but Proof takes a string. would be nice for the two to be consistent.

**Review by mvdan — APPROVED** [review]

**Comment by p4u** [comment]
Reduced Proof to 1 byte and added envelopeType of 1 byte.

Up to 256 kind of proofs should be enough.

@mvdan can you review again? specially the binary operations which I tried to make as safe as possible.

**Pull request merged**